kernel/src/scan/mod.rs

```diff
@@ -92,25 +96,169 @@ impl ScanBuilder {
 
-        // important! before a read/write to the table we must check it is supported
-        self.snapshot.protocol().ensure_read_supported()?;
+        // If we have a predicate, verify the columns it references and apply column mapping. First,
+        // get the set of references; use that to filter the schema to only the columns of interest
+        // (and verify that all referenced columns exist); then use the resulting logical/physical
+        // mappings to rewrite the expression with physical column names.
+        //
+        // NOTE: It is possible the predicate doesn't reference any schema; in that case it must
+        // consist entirely of literal expressions. We can't just evaluate it directly, because data
+        // skipping has specific rules about what kinds of expressions it supports. So just
+        // propagate the empty schema and let the skipping machinery deal with it.
+        let physical_predicate = match self.predicate {
+            Some(predicate) => Self::build_physical_predicate(predicate, &logical_schema)?,
+            None => PhysicalPredicate::None,
+        };
 
         Ok(Scan {
             snapshot: self.snapshot,
             logical_schema,
-            physical_schema,
-            predicate: self.predicate,
-            all_fields,
-            have_partition_cols,
+            physical_schema: Arc::new(StructType::new(state_info.read_fields)),
+            physical_predicate,
+            all_fields: state_info.all_fields,
+            have_partition_cols: state_info.have_partition_cols,
         })
     }
+
+    fn build_physical_predicate(
+        predicate: ExpressionRef,
+        logical_schema: &Schema,
+    ) -> DeltaResult<PhysicalPredicate> {
+        if can_statically_skip_all_files(&predicate) {
+            return Ok(PhysicalPredicate::StaticSkipAll);
+        }
+        let mut get_referenced_fields = GetReferencedFields {
+            references: predicate.references(),
+            column_mappings: HashMap::new(),
+            logical_path: vec![],
+            physical_path: vec![],
+        };
+        let schema_opt = get_referenced_fields.transform_struct(logical_schema);
+        if let Some(missing_column) = get_referenced_fields.references.into_iter().next() {
+            // NOTE: It's a pretty serious engine bug if we got this far with a query whose WHERE
+            // clause has invalid column references. Data skipping is best-effort and the predicate
+            // anyway needs to be evaluated against every row of data -- which is impossible if the
+            // columns are missing/invalid. Just blow up instead of trying to handle it gracefully.
+            return Err(Error::missing_column(format!(
+                "Predicate references unknown column: {missing_column}"
+            )));
+        }
+        let Some(schema) = schema_opt else {
+            // We couldn't statically skip all files, so this predicate is useless
```

Comment:
this comment confused me :) I think you mean 'We can't use the stats to skip any files, so this predicate is useless"

Reply:
The predicate doesn't statically skip all files, and it doesn't reference any columns that could dynamically change its behavior, so it's useless.

Reply:
Right, that's an even better way of saying it (not referencing the columns being the _cause_ of not being able to use the stats)